From 77e130c1e2f2879ffe9b7add0873101c50208d1a Mon Sep 17 00:00:00 2001
From: Ben Gamari <ben@smart-cactus.org>
Date: Mon, 2 Oct 2017 09:45:05 -0400
Subject: [PATCH] Ensure libffi.map is generated

---
 configure.ac | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/configure.ac b/configure.ac
index d155a6d..f3162d5 100644
--- a/configure.ac
+++ b/configure.ac
@@ -383,6 +383,6 @@ test -d src/$TARGETDIR || mkdir src/$TARGETDIR
 
 AC_CONFIG_LINKS(include/ffitarget.h:src/$TARGETDIR/ffitarget.h)
 
-AC_CONFIG_FILES(include/Makefile include/ffi.h Makefile testsuite/Makefile man/Makefile doc/Makefile libffi.pc)
+AC_CONFIG_FILES(include/Makefile include/ffi.h Makefile testsuite/Makefile man/Makefile doc/Makefile libffi.pc libffi.map)
 
 AC_OUTPUT
-- 
2.11.0

